### ISSUE TITLE
As a smith I want to investigate how to avoid editor functionality to be blocked by the LSP builder

### ISSUE THREAD

**Issue by mor-n4** [eclipse/n4js#1768]
This user story is about investigating ways of making editor functionality in the LSP server (e.g. hover, hyperlinks, content assist) independent from the builder (as far as possible). A final implementation of the solution is not part of this task.

#### Background

Currently all operations in the LSP server are managed by class `XRequestManager` as "requests", either a "read request" or a "write request". So far, write requests are used only for the incremental builder; read requests are used for all other functionality (e.g. hover, hyperlinks, content assist, code actions, commands).

As a rule implemented in `XRequestManager`, incoming write-requests cancel all running/pending requests (both read and write, except for the non-cancelable part of write requests). Incoming read requests do not cancel anything; they will wait for a running write request to complete.

#### Details

This means that for as long as an incremental build (i.e. write request) is running, all non-build functionality (i.e. read requests) is unavailable, for example, hovers, hyperlinks, content assist. Interestingly, this is true even for the "N4JS: Rebuild" command, because before that command has a chance to trigger a clean and full build (which might cancel a running build) it is first wrapped in a read request.

Since the incremental build triggered by a change in an opened file (saving not required) will not only rebuild this one file but also all affected resources across the entire code base, this can take 10sec or more in large projects.
